sdk/servicebus/Microsoft.Azure.WebJobs.Extensions.ServiceBus/src/Listeners/ServiceBusTargetScaler.cs

```diff
@@ -0,0 +1,83 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+using Azure.Messaging.ServiceBus;
+using Azure.Messaging.ServiceBus.Administration;
+using Microsoft.Azure.WebJobs.Extensions.ServiceBus.Config;
+using Microsoft.Azure.WebJobs.Host.Scale;
+using Microsoft.Azure.WebJobs.ServiceBus;
+using Microsoft.Azure.WebJobs.ServiceBus.Listeners;
+using Microsoft.Extensions.Logging;
+using System;
+using System.Globalization;
+using System.Threading.Tasks;
+
+namespace Microsoft.Azure.WebJobs.Extensions.ServiceBus.Listeners
+{
+    public class ServiceBusTargetScaler : ITargetScaler
+    {
+        private readonly string _functionId;
+        private readonly ServiceBusMetricsProvider _serviceBusMetricsProvider;
+        private readonly ServiceBusOptions _options;
+        private readonly bool _isSessionsEnabled;
+        private readonly TargetScalerDescriptor _targetScalerDescriptor;
+        private readonly string _entityPath;
+        private readonly ILogger _logger;
+
+        public ServiceBusTargetScaler(
+            string functionId,
+            string entityPath,
+            ServiceBusEntityType entityType,
+            Lazy<ServiceBusReceiver> receiver,
+            Lazy<ServiceBusAdministrationClient> administrationClient,
+            ServiceBusOptions options,
+            bool isSessionsEnabled,
+            ILoggerFactory loggerFactory
+            )
+        {
+            _functionId = functionId;
+            _serviceBusMetricsProvider = new ServiceBusMetricsProvider(entityPath, entityType, receiver, administrationClient, loggerFactory);
+            _entityPath = entityPath;
+            _targetScalerDescriptor = new TargetScalerDescriptor(functionId);
+            _logger = loggerFactory.CreateLogger<ServiceBusTargetScaler>();
+            _options = options;
+            _isSessionsEnabled = isSessionsEnabled;
+        }
+
+        public TargetScalerDescriptor TargetScalerDescriptor => _targetScalerDescriptor;
+
+        public async Task<TargetScalerResult> GetScaleResultAsync(TargetScalerContext context)
+        {
+            ServiceBusTriggerMetrics metrics = await _serviceBusMetricsProvider.GetMetricsAsync(true).ConfigureAwait(false);
+            return GetScaleResultInternal(context, metrics.MessageCount);
```

Comment:
Target scaler only needs this MessageCount, doesn't need QueueTime or anything else. So calling GetMetricsAsync will result in unnecessary Peek operations to determine QueueTime when it's not even used. So assuming we also need PartitionCount, I think you'd have a separate method ServiceBusMetricsProvider.GetRuntimeProperties that returns a type with MessageCount/PartitionCount. And that method would be used here, as well as from GetQueueMetricsAsync/GetTopicMetricsAsync.

Reply:
there are 3 await operations in the `ServiceBusMetricsProvider`:
1. `PeekMessageAsync`
2. `GetQueueRuntimePropertiesAsync`or `GetSubscriptionRuntimePropertiesAsync`
3. `GetQueueAsync` or `GetTopicAsync`

We need 1 and 2 to get active message count. PeekMessageAsync is needed for ['no manage claim case'](https://github.com/Azure/azure-sdk-for-net/pull/31367/files/8110f7a257d6506a7906bc07b6284998633f3d76#diff-a3f037667869ab7bb60b69537d01e7999b070bc2318b0478ead0d987bc4000c9R182)